In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report, recall_score

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, VotingClassifier

In [5]:
df = pd.read_csv("novagen_dataset.csv")

X = df.drop("Target",axis=1)
y = df["Target"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42,stratify=y
)

In [7]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [10]:
# Logistic regression (with regularization)
log_reg = LogisticRegression(
    penalty="l2",
    solver="liblinear",
    max_iter=1000
)

log_reg.fit(X_train_scaled, y_train)

y_pred_lr = log_reg.predict(X_test_scaled)

print("logistic regression accuracy: ", accuracy_score(y_test, y_pred_lr))
print("logistic regression recall: ", recall_score(y_test, y_pred_lr))
print("classification report: ", classification_report(y_test, y_pred_lr))

logistic regression accuracy:  0.8141361256544503
logistic regression recall:  0.8283132530120482
classification report:                precision    recall  f1-score   support

           0       0.81      0.80      0.80       914
           1       0.82      0.83      0.82       996

    accuracy                           0.81      1910
   macro avg       0.81      0.81      0.81      1910
weighted avg       0.81      0.81      0.81      1910



In [12]:
knn = KNeighborsClassifier(
    n_neighbors=5,
    metric="euclidean"
)

knn.fit(X_train_scaled, y_train)

y_pred_knn = knn.predict(X_test_scaled)

print("logistic regression accuracy: ", accuracy_score(y_test, y_pred_lr))
print("logistic regression recall: ", recall_score(y_test, y_pred_lr))
print("classification report: ", classification_report(y_test, y_pred_lr))

logistic regression accuracy:  0.8141361256544503
logistic regression recall:  0.8283132530120482
classification report:                precision    recall  f1-score   support

           0       0.81      0.80      0.80       914
           1       0.82      0.83      0.82       996

    accuracy                           0.81      1910
   macro avg       0.81      0.81      0.81      1910
weighted avg       0.81      0.81      0.81      1910



In [13]:
rf = RandomForestClassifier(
    n_estimators=200,
    max_depth=None,
    random_state=42
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

print("logistic regression accuracy: ", accuracy_score(y_test, y_pred_rf))
print("logistic regression recall: ", recall_score(y_test, y_pred_rf))
print("classification report: ", classification_report(y_test, y_pred_rf))

logistic regression accuracy:  0.937696335078534
logistic regression recall:  0.9588353413654619
classification report:                precision    recall  f1-score   support

           0       0.95      0.91      0.93       914
           1       0.92      0.96      0.94       996

    accuracy                           0.94      1910
   macro avg       0.94      0.94      0.94      1910
weighted avg       0.94      0.94      0.94      1910



In [14]:
gb = GradientBoostingClassifier(
    n_estimators=150,
    learning_rate=0.1,
    random_state=12
)

gb.fit(X_train, y_train)

y_pred_gb = gb.predict(X_test)

print("logistic regression accuracy: ", accuracy_score(y_test, y_pred_gb))
print("logistic regression recall: ", recall_score(y_test, y_pred_gb))
print("classification report: ", classification_report(y_test, y_pred_gb))

logistic regression accuracy:  0.930890052356021
logistic regression recall:  0.9497991967871486
classification report:                precision    recall  f1-score   support

           0       0.94      0.91      0.93       914
           1       0.92      0.95      0.93       996

    accuracy                           0.93      1910
   macro avg       0.93      0.93      0.93      1910
weighted avg       0.93      0.93      0.93      1910



In [18]:
voting_clf = VotingClassifier(
    estimators = [
        ("lr",LogisticRegression(max_iter=1000, solver="liblinear")),
        ("knn",KNeighborsClassifier(n_neighbors=5)),
        ("rf",RandomForestClassifier(n_estimators=200, random_state=42))
    ],
    voting="soft"
)

voting_clf.fit(X_train_scaled, y_train)

y_pred_vote = voting_clf.predict(X_test_scaled)

print("logistic regression accuracy: ", accuracy_score(y_test, y_pred_vote))
print("logistic regression recall: ", recall_score(y_test, y_pred_vote))
print("classification report: ", classification_report(y_test, y_pred_vote))

logistic regression accuracy:  0.9162303664921466
logistic regression recall:  0.9307228915662651
classification report:                precision    recall  f1-score   support

           0       0.92      0.90      0.91       914
           1       0.91      0.93      0.92       996

    accuracy                           0.92      1910
   macro avg       0.92      0.92      0.92      1910
weighted avg       0.92      0.92      0.92      1910



## Results

| Model                | Recall |
|----------------------|:------:|
| Logistic Regression  | 82.8%  |
| KNN                  | 88.3%  |
| Random Forest        | 95.8%  |
| Gradient Boosting    | 94.9%  |
| Voting Classifier    | 93.07% |

### Best Classifier that we should use for NovaGen(based on Recall) - Random Forest with accuracy of 93.7%